lib/hyrax/transactions/transaction.rb

```diff
@@ -0,0 +1,141 @@
+    #   change_set.title = ['comet in moominland']
+    #
+    #   tx.call(change_set) # => Success(#<Hyrax::Work ...>)
+    #
+    # @example with a failure
+    #   class ChangeSetWithTitleValidation < Hyrax::ChangeSet
+    #     self.fields = [:title]
+    #     validates :title, presence: true
+    #   end
+    #
+    #   change_set = ChangeSetWithTitleValidation.new(Hyrax::Work.new)
+    #   change_set.title = []
+    #
+    #   tx.call(change_set)
+    #   # => Failure([:failed_validation, #<Reform::Form::ActiveModel::Errors ...>]
+    #
+    # @example unwapping values safely with handling for failures
+    #   tx.call(change_set).value_or { |failure| "uh oh: #{failure} }
+    #
+    # @example a pattern for subclassing to create new transactions
+    #   class CustomTransaction < Transaction
+    #     DEFAULT_STEPS = ['step.1', 'step.2', 'step.3'].freeze
+    #
+    #     def initialize(container: Container, steps: DEFAULT_STEPS)
+    #       super
+    #     end
+    #   end
+    #
+    #   tx = CustomTransaction.new
+    #   tx.call(:some_value)
+    #
+    class Transaction
+      include Dry::Monads[:result]
+      include Dry::Monads::Do.for(:call)
+
+      ##
+      # @!attribute [rw] container
+      #   @return [Container]
+      # @!attribute [rw] steps
+      #   @return [Array<String>]
+      attr_accessor :container, :steps
+
+      ##
+      # @api public
+      #
+      # @param [Container] container
+      # @param [Array<String>] steps
+      def initialize(container: Container, steps:)
+        self.container = container
+        self.steps     = steps
+      end
+
+      ##
+      # @api public
+      #
+      # Run each step name in `#steps` by resolving the name in the given
+      # `container` and passing a value to call. Each step must return a
+      # `Result`. In the event of a `Success`, the wrapped value is passed
+      # through to the next step; for `Failure` the whole chain is short-
+      # circuited and the failure result is given.
+      #
+      # @param [Object] value
+      #
+      # @return [Dry::Monads::Result] either the `Success` result of the entire
+      #   `#steps` chain over `value`, or a `Failure` from the failed step.
+      # @see Dry::Monads::Do
+      def call(value)
+        Success(
+          steps.inject(value) do |val, step_name|
+            yield container[step_name].call(val, *step_arguments_for(step_name))
+          end
+        )
+      end
+
+      ##
+      # @param [Hash<Object, Array>] args
+      #
+      # @return [Transaction] returns self
+      def with_step_args(args)
+        raise(ArgumentError, key_err_msg(args.keys)) if
+          args.keys.any? { |key| !step?(key) }
+
+        @_step_args = args
+        self
+      end
+
+      private
+
+        ##
+        # @api private
+        # @param [String, Symbol] step_name
+        # @return [Array]
+        def step_arguments_for(step_name)
+          Array.wrap((@_step_args || {})[step_name])
```

Comment:
There is a bit of visual gymnastics with this one; It is hard to scan like a medium complexity regexp. Perhaps separate the hash lookup/setting:

```ruby
step_args = @_set_args || {}
Array.wrap(step_args[step_name])
```